tests/conftest.py

```diff
@@ -69,11 +48,10 @@ def docker() -> DockerClient:  # type: ignore[misc]  # No docker types.
 
 
 @pytest.fixture(scope='session')
-def redis_server(  # type: ignore[misc]  # No docker types.
+async def redis_server(  # type: ignore[misc]  # No docker types.
```

Comment:
This change is likely the reason of the problem — pytest fixtures are synchronous AFAIK.

Reply:
Hmm, this suggested that fixtures could be async:
https://stackoverflow.com/questions/49936724/async-fixtures-with-pytest#49940520

Reply:
I think it may depend on what pytest plugins get used. Anyway, this should give you an idea of where to start bisecting the issue.

Reply:
Fixtures that pytest-aiohttp provides usually rely on having a disposable event loop where those tests will run. If a fixture is a coroutine on the outer pytest level, I suspect it may run in a different event loop and I'm not ready to guess what the consequences of that would be.

Reply:
FWIW try to make this work w/o this part of the refactoring and see if that helps.

Reply:
It's not refactoring, it's just trying to get the thing working. :P
I'll revert these changes and take another look at the previous errors.

Reply:
Yep, so now back to `There is no current event loop in thread 'MainThread'.`
As the `loop` parameter has been removed from just about everything now, I don't know how to make this work.
I'm not sure why `loop.run_until_complete()` isn't able to set itself as the current event loop for the length of the run.

Reply:
I tried changing `asyncio.set_event_loop(loop)` instead of `None`. But, that hasn't made a difference.

Reply:
I think that maybe when there's `loop.run_until_complete(do_smth())`, that `do_smth()` triggers the problem. Maybe try wrapping it into another coroutine and see what happens?
As in
```python
async def coro():
    return await do_smth()

loop.run_until_complete(coro())
```

Reply:
Try running just one test with `--pdb` maybe? Or inject a `breakpoint()` right before?

Reply:
Didn't help. I don't have the docker stuff setup to run the full tests locally, but I commented out the docker parts and have just found that it successfully runs past the point of failure in CI. So, I can't even reproduce the exception locally...

Reply:
I'm feeling like this is something @asvetlov would be able to figure out?

Reply:
My 2 cents: The loop fixture sets the event loop to `None` in its cleanup [here](https://github.com/aio-libs/aiohttp-session/blob/1a589e37371c35b52d637035e54827db6a175c55/tests/conftest.py#L52) and setup [here](https://github.com/aio-libs/aiohttp-session/blob/1a589e37371c35b52d637035e54827db6a175c55/tests/conftest.py#L43). Reading the error, `aioredis` tries to acquire an `asyncio.Lock` which in turn internally runs `events.get_event_loop()` which since the `loop` fixture set the running loop to `None` fails.

Suggestion:
try setting the global loop to the fixture loop inside the redis_server fixture (i.e. add [here](https://github.com/aio-libs/aiohttp-session/blob/1a589e37371c35b52d637035e54827db6a175c55/tests/conftest.py#L74) `asyncio.set_event_loop(loop)`). This is bad practice IMO (so it wouldn't be a permanent solution) but if this error is then fixed it will tell us we are at least in a good path.

Sorry for not testing it myself ATM but I only have python 3.8 on my work machine (and the tests already fail on 3.8). I'll try to offer a bit more hands-on help during the weekend (but tentative if I'll have time).

Reply:
Also, [this](https://aioredis.readthedocs.io/en/latest/migration/#connecting-to-redis) suggests that `aioredis.from_url` is not a coro.

Reply:
Minimum reproducible example:

```py
>>> import asyncio
>>> import aioredis
>>> r = aioredis.from_url("redis://{}:{}".format("foo", 123))
>>> asyncio.set_event_loop(None)
>>> r = aioredis.from_url("redis://{}:{}".format("foo", 123))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/panagiks/.venvs/aioredis2/lib/python3.8/site-packages/aioredis/utils.py", line 25, in from_url
    return Redis.from_url(url, **kwargs)
  File "/home/panagiks/.venvs/aioredis2/lib/python3.8/site-packages/aioredis/client.py", line 814, in from_url
    connection_pool = ConnectionPool.from_url(url, **kwargs)
  File "/home/panagiks/.venvs/aioredis2/lib/python3.8/site-packages/aioredis/connection.py", line 1245, in from_url
    return cls(**kwargs)
  File "/home/panagiks/.venvs/aioredis2/lib/python3.8/site-packages/aioredis/connection.py", line 1270, in __init__
    self._lock = asyncio.Lock()
  File "/usr/lib/python3.8/asyncio/locks.py", line 164, in __init__
    self._loop = events.get_event_loop()
  File "/usr/lib/python3.8/asyncio/events.py", line 639, in get_event_loop
    raise RuntimeError('There is no current event loop in thread %r.'
RuntimeError: There is no current event loop in thread 'MainThread'.
```

Reply:
As for why the loop in not set inside `loop.run_until_complete` the answer is because it is never executed. Since `aioredis.from_url` is a synchronous function, it is executed before `loop.run_until_complete` and its result would have been passed to it as an argument (like running `loop.run_until_complete(str(42))` the `str(42)` would be executed first and then `loop.run_until_complete("42")` would run). Sorry for the multi-message spam :( (new ideas coming in just as I hit 'Comment') That's my last message for now :D

Reply:
Thanks for looking at this.

> try setting the global loop to the fixture loop inside the redis_server fixture
I already tried setting the loop in the loop fixture in a previous commit. I've tried setting it in the redis_server as you suggested, and I think it's working! We've got some new errors now, so I'll try and figure those out now.

> Sorry for not testing it myself ATM but I only have python 3.8 on my work machine (and the tests already fail on 3.8).
Yep, this is first step towards getting 3.8+ running.

> Also, this suggests that aioredis.from_url is not a coro.
Yes, seems a little weird. It returns the client, then an `await`, `async with` or simply executing a command all do the same thing and initialise the client.

Reply:
OK, looks like it's working now. If you've got any ideas on a better way to do it, that would be great. It looks like we have the same issue with aiomcache when we remove the loop parameter from that.

We are also getting connection errors on about half the CI runs, requiring them to be rerun repeatedly. Any ideas to fix that would be great as well:
`aioredis.exceptions.ConnectionError: Error 111 connecting to 172.17.0.3:6379. 111.`

Reply:
Well, maybe implement an exponential backoff? Or a fixture that would wait for the service to be actually up and running before proceeding with tests?

Reply:
I'll follow up these changes in another PR. @panagiks Let me know if you come up with any improvements.